### ISSUE TITLE
MetaMask performance got worse in v8

### ISSUE THREAD

**Issue by wjmelements** [MetaMask/metamask-extension#8991]
Reopening https://github.com/MetaMask/metamask-extension/issues/8572 for version 8.0.4

Whether connecting, opening the popup, or sending a transaction, I am met by a blank page that must be watched the whole time or all progress will be lost.

From clicking the fox icon, it takes 6 seconds before the blank white window appears and 19 seconds before I can see an account.

**Comment by rekmarks** [comment]
@wjmelements thank you for alerting us to this.

@brad-decker, are there any low-hanging fruits in the UI that could help with this, like buffering the transaction list?

**Comment by wjmelements** [comment]
Is there something I can do to help you profile the delay?

**Comment by bunsen2** [comment]
Same here.  New version is running at least 10x slower than the old version.

**Comment by Gudahtt** [comment]
@wjmelements To start, we'll need to reproduce this delay you're experiencing. Are you only experiencing this when using an account with >1000 transactions, as with #8572? Which browser are you using?

**Comment by wjmelements** [comment]
@Gudahtt Unlike #8572 I also experience a delay when I have an unused account selected, but it could still be relevant that I have two accounts with over 2000 transactions and one with over 5000.

I am using Chrome.

**Comment by wjmelements** [comment]
I profiled the background process during the delay using the chrome devtools, but it looks like minified garbage. Might still be helpful.
<img width="1434" alt="Screen Shot 2020-07-14 at 7 49 48 PM" src="https://user-images.githubusercontent.com/799573/87497805-2f604680-c60b-11ea-8658-916d190eb682.png">

**Comment by wjmelements** [comment]
This looks relevant; lots of GC, with write() taking the most time
<img width="1434" alt="Screen Shot 2020-07-14 at 7 52 14 PM" src="https://user-images.githubusercontent.com/799573/87497955-86661b80-c60b-11ea-9947-7f9ef31037f7.png">

**Comment by rstormsf** [comment]
![image](https://user-images.githubusercontent.com/9360827/87502687-11004800-c617-11ea-80b2-84504780923d.png)
speed up is also broken in queued tx

**Comment by danfinlay** [comment]
@wjmelements A few questions:

- Are these thousands of txs incoming, or are outgoing also listed?
- To clarify, this is only on v8, even though you opened the issue before that was live? (I assume you were living off `develop`?)

I ask because [we have long limited the length of the outgoing history](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/transactions/tx-state-manager.js#L156), so either this is a bug and we're showing [more than 40](https://github.com/MetaMask/metamask-extension/blob/4f0a20536903554b228cf4df00de6feb3c6063d5/app/scripts/metamask-controller.js#L238) or this is related to how v8 now shows incoming txs and does not impose that limit.

Would we solve this issue if we truncated the incoming tx list for now, until we allow a more incremental loading of tx history?

**Comment by ElderOrb** [comment]
I'm experiencing this issue too

**Comment by whymarrh** [comment]
@ElderOrb would you be able to share a rough count of how many transactions you have across your accounts?

**Comment by whymarrh** [comment]
@danfinlay tx performance was bad in 7.x as well (i.e. #8572) and v8 has possibly made it worse. I don't believe this is related to the tx state history, but I'll look at that.

@wjmelements thanks for re-opening this, I'm looking into it now.

**Comment by bunsen2** [comment]
> Same here. New version is running at least 10x slower than the old version.

It's running smoothly after I deleted the extension and re-installed it

**Comment by wjmelements** [comment]
> @wjmelements A few questions:
> 
> * Are these thousands of txs incoming, or are outgoing also listed?

Outgoing

> * To clarify, this is only on v8, even though you opened the issue before that was live? (I assume you were living off `develop`?)

The prior issue was for v7.

**Comment by whymarrh** [comment]
I can confirm that this is a severe performance problem. The "good" news is that this isn't _significantly worse_ in v8—the slowness here is a result of the serializing/deserializing of state between MetaMask's background (persistent) and UI processes, and this hasn't changed recently.

I have a stopgap fix (#9010) that will limit the number of transactions exposed to the UI (more or less). This does indeed increase the responsiveness of the application quite drastically. 2.5k-3k transactions can take upwards of 4.5 seconds to move around internally, and with #9010 that is down to 90ms for de/serialization in some tests.

<details>
<summary>Before vs. after profile screenshots</summary>
<img width="1526" alt="" src="https://user-images.githubusercontent.com/1623628/87613203-36f51d80-c6e7-11ea-89bc-11a1cc2f3b1e.png">
<img width="1526" alt="" src="https://user-images.githubusercontent.com/1623628/87613215-3bb9d180-c6e7-11ea-8d85-aff3acbd0374.png">
</details>

The stopgap fix is just that. The bad news is that we don't have a _good_ short-term fix. A proper fix for this will involve reworking how our background and UI contexts communicate, which is definitely on our radar, but isn't something we've got planned for a patch release.

> It's running smoothly after I deleted the extension and re-installed it

A _very immediate_ workaround is, yes, to restore from a seed phrase, which will reset the installation's tracked transactions. This isn't ideal, obviously, but is the only option until 8.0.5 is published.

**Comment by xmaysonnave** [comment]
Metamask v8.0.4, debian buster, I5-7300U, 16GB, NVME

Here some informations how slow could be the UI.

I started to use Metamask encryption.

1 - Encryption performed with eth-sig-util and the public encryption key retrieved from Metamask
The sample is a 3MB sample (base64)
```
sigUtil.encrypt(
  currentPublicKey,
  { data: text },
  'x25519-xsalsa20-poly1305'
)
```
Here is my output:
`Ethereum Encrypt: 152ms, In: 2524460, Out: 3366129, Ratio: 133%`

2 - Decryption occurs in Metamask
Metamask is unlocked and the DApp has already the permission to access Metamask
Here is my output:
`Ethereum Decrypt: 24435ms, In: 3366157, Out: 2524480, Ratio: 74%`
In some situation, Metamask when locked and no permission the UI could take up to 50s.
```
var tStart = new Date();
var decryptedText = await provider.request({ method: "eth_decrypt", params: [text, accounts[0]] })
if (decryptedText !== undefined || decryptedText !== null) {
  var tStop = new Date()-tStart;
  var ratio = Math.floor(decryptedText.length*100/text.length);
  console.log(`Ethereum Decrypt: ${tStop}ms, In: ${text.length}, Out: ${decryptedText.length}, Ratio: ${ratio}%`);
}
```
3 - Decryption when tested with only eth-sig-util and the appropriate private key. 
It takes roughly 300ms to decrypt the content.

Thanks

**Comment by Gudahtt** [comment]
The stopgap measure implemented in #9010 has hopefully worked well as a temporary solution, but I'm going to leave this open for now until we've at least documented in GitHub what our longer-term solution for this will be. In the meantime I'll mark this as "blocked", since further work is blocked on these yet-to-be-specified improvements to our background <=> UI communication.
